### ISSUE TITLE
Constraint: escalate fails when depends_on jana2 +zmq due to cppzmq py-pyzmq constraints

### ISSUE THREAD

**Issue by wdconinc** [eic/eic-spack#55]
### Steps to reproduce the issue

Make sure escalate and ejana include jana2 +zmq as a dependency.
```console
spack spec escalate@develop
```
Returns
```
[wdconinc@cvmfswrite01 ~/eic-spack]$ spack spec -I escalate@develop target=x86_64
Input spec
--------------------------------
 -   escalate@develop arch=linux-None-x86_64

Concretized
--------------------------------
==> Error: An unsatisfiable version constraint has been detected for spec:

    libzmq@4.3.2%gcc@9.3.0+libsodium arch=linux-rhel7-x86_64
        ^libsodium
        ^pkgconf@1.7.3%gcc@9.3.0 arch=linux-rhel7-x86_64


while trying to concretize the partial spec:

    cppzmq@4.3.0%gcc@9.3.0 build_type=RelWithDebInfo arch=linux-rhel7-x86_64


cppzmq requires libzmq version 4.2.5, but spec asked for 4.3.2
```

### Information on your system

```
[wdconinc@cvmfswrite01 ~/eic-spack]$ spack debug report
* **Spack:** 0.14.2-1720-6a53c5f97
* **Python:** 2.7.5
* **Platform:** linux-rhel7-skylake_avx512
```

### Potential resolution

At the core, this seems to be caused by jana2 which requires cppzmq (most recent @4.3.0) which requires libzmq@4.2.5 exactly, while escalate also includes py-jupyter which requires py-pyzmq (most recent @18.1.0) which requires libzmq (most recent @4.3.2). So, resolution fails because cppzmq wants a lower version than the most recent version required by py-pyzmq, and the concretization does not resolve to use libzmq@4.2.5 for some reason.

Unfortunately the whole point of ejana is kinda to have both jana2 and py-jupyter enabled. Currently this excludes any zmq functionality in jana2.

Possibly solution is to use a less strict cppzmq requirement, e.g. depends_on('libzmq@4.2.5:4.2.99'), but that needs testing.

**Comment by wdconinc** [comment]
This will probably require an upstream update to cppzmq to add newer versions, to add open upper-ended dependencies on libzmq, then a bunch of testing.

**Comment by wdconinc** [comment]
In the meantime: `spack spec -t escalate@develop target=x86_64 ^libzmq@4.2.5` as a workaround.

**Comment by wdconinc** [comment]
Also, simultaneously we might have to submit a bug report to spack to fix the dependency resolution since it should have been able to resolve this.

**Comment by wdconinc** [comment]
https://github.com/spack/spack/issues/18086

https://github.com/spack/spack/issues/18087

**Comment by wdconinc** [comment]
More recent versions of cppzmq now in upstream/develop. Waiting for v0.16 to resolve this.

**Comment by wdconinc** [comment]
This has been resolved in develop and v0.18.
